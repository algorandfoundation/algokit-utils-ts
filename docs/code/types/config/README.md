[**@algorandfoundation/algokit-utils**](../../README.md)

***

[@algorandfoundation/algokit-utils](../../README.md) / types/config

# types/config

## Classes

- [UpdatableConfig](classes/UpdatableConfig.md)

## Interfaces

- [Config](interfaces/Config.md)
