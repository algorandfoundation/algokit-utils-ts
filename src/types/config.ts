import { AsyncEventEmitter } from './async-event-emitter'
import { Logger, consoleLogger, nullLogger } from './logging'

/** The AlgoKit configuration type */
export interface Config {
  /** Logger */
  logger: Logger
  /** Whether or not debug mode is enabled */
  debug: boolean
  /** The path to the project root directory */
  projectRoot: string | null
  /** Indicates whether to trace all operations */
  traceAll: boolean
  /** The size of the trace buffer in megabytes */
  traceBufferSizeMb: number
  /** The maximum depth to search for a specific file */
  maxSearchDepth: number
  /**
   * **WARNING**: This is not production-ready due incompatability with rekeyed
   * accounts and simulate. This will eventually be enabled by default once
   * [this issue](https://github.com/algorand/go-algorand/issues/5914) is closed.
   *
   * Whether to enable populateAppCallResources in sendParams by default.
   * Default value is false.
   */
  populateAppCallResources: boolean

  events: AsyncEventEmitter
}

/** Updatable AlgoKit config */
export class UpdatableConfig implements Readonly<Config> {
  private config: Config

  get populateAppCallResources() {
    return this.config.populateAppCallResources
  }

  get logger() {
    return this.config.logger
  }

  get debug() {
    return this.config.debug
  }

  get projectRoot() {
    return this.config.projectRoot
  }

  get traceAll() {
    return this.config.traceAll
  }

  get traceBufferSizeMb() {
    return this.config.traceBufferSizeMb
  }

  get maxSearchDepth() {
    return this.config.maxSearchDepth
  }

  get events() {
    return this.config.events
  }

  /**
   * Returns the current logger, or the null logger if true is passed in to `returnNullLogger`
   * @param returnNullLogger Whether or not to return the null logger
   * @returns The requested logger
   */
  getLogger(returnNullLogger?: boolean) {
    if (returnNullLogger) {
      return nullLogger
    }

    return this.logger
  }

  /**
   * Temporarily run with debug set to true.
   * @param lambda A lambda expression with code to run with debug config set to true
   */
  withDebug(lambda: () => unknown) {
    const original = this.config.debug
    try {
      this.config.debug = true
      lambda()
    } finally {
      this.config.debug = original
    }
  }

  constructor() {
    this.config = {
      logger: consoleLogger,
      debug: false,
      projectRoot: null,
      traceAll: false,
      traceBufferSizeMb: 256,
      maxSearchDepth: 10,
      populateAppCallResources: false,
      events: new AsyncEventEmitter(),
    }
  }

  /**
   * Update the AlgoKit configuration with your own configuration settings
   * @param newConfig Partial or complete config to replace
   */
  configure(newConfig: Partial<Config>) {
    this.config = { ...this.config, ...newConfig }
  }
}
