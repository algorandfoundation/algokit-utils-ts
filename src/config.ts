import { UpdatableConfig } from './types/config'

/** The AlgoKit config. To update it use the configure method. */
export const Config = new UpdatableConfig()
