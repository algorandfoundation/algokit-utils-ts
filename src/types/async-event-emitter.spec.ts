/* eslint-disable no-console */
import { Config } from '../config'
import { algoKitLogCaptureFixture } from '../testing/fixtures/algokit-log-capture-fixture'

describe('async-event-emitter', () => {
  const logging = algoKitLogCaptureFixture()
  beforeEach(logging.beforeEach)
  afterEach(logging.afterEach)

  const mockRegisterNodeDebugHandlers = () => {
    // TODO: Define types for the events
    // TODO: Question - Would this be the package we'd use to surface the UI widget for downloading a simulate trace?
    Config.events.on('event_a', (data: unknown) => {
      logging.testLogger.info('event_a', data)
    })
  }

  it('users code would looks like this', async () => {
    Config.configure({
      debug: true,
    })
    mockRegisterNodeDebugHandlers() // imitates invocation of algokit-utils-ts-debug

    Config.events.emitAsync('event_a', {
      hello: 'world',
    })

    // await for the event to be processed by checking if capturedLogs is empty
    while (logging.testLogger.capturedLogs.length === 0) {
      await new Promise((resolve) => setTimeout(resolve, 100))
    }

    expect(logging.testLogger.capturedLogs).toMatchSnapshot()
  })
})
