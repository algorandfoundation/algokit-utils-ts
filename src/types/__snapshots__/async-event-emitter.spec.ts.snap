// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`async-event-emitter > users code would looks like this 1`] = `
[
  "INFO: event_a | [{"hello":"world"}]",
]
`;
